syntax="proto3";

package clone.hardware_driver;

message SendDirectMessage {
  uint32 node_id = 1;
  uint32 valves = 2;
}

message SendManyDirectMessage {
  map<uint32, uint32> data = 1;
}

message ImpulseControl {
  uint32 valves = 1;
  uint32 delay = 2;
}

message SendImpulseMessage {
  uint32 node_id = 1;
  ImpulseControl control = 2;
}

message SendManyImpulseMessage {
  map<uint32, ImpulseControl> data = 1;
}

message SendPulseMessage {
  uint32 node_id = 1;
  uint32 valve_id = 2;
  uint32 pulse_len_ms = 3;
  uint32 delay_len_ms = 4;
  uint32 duration_ms = 5;
}

message PressureControl {
  repeated int32 pressures = 1;
}

message SendManyPressureMessage {
  map<uint32, PressureControl> data = 1;
}

message PinchValveControl {
  enum PositionsType {
      POSITION_UNKNOWN = 0;	// Cam position unknown
      BOTH_CLOSED = 1;	// Both inlet and outlet are closed
      BOTH_OPENED = 2;	// Both inlet and outlet are opened, AKA flush
      INLET_FULLY_OPENED = 3;	// Inlet valve fully opened
      INLET_PARTIALLY_OPENED = 4;	// Inlet valve opened, but only partially
      OUTLET_FULLY_OPENED = 5;	// Outlet valve fully opened
      OUTLET_PARTIALLY_OPENED = 6;	// Outlet valve opened, but only partially
  }

  enum ControlMode {
      RESERVED = 0;
      ANGLE = 1;
      PRESSURE = 2;
      POSITIONS = 3;
  }

  ControlMode mode = 1;
  int32 value = 2;
}

message SendPinchValveControlMessage {
  uint32 node_id = 1;
  PinchValveControl control = 2;
}

message SendManyPinchValveControlMessage {
  map<uint32, PinchValveControl> data = 1;
}

message GetNodesMessage {
  // bool rediscover = 1;
  optional uint32 product_id = 2;
}

message BusDevice {
  uint32 node_id = 1;
  uint32 product_id = 2;
}

message NodeList {
  repeated BusDevice nodes = 1;
}

message NodeMap {
  map<string, NodeList> nodes = 1;
}

enum PinchValveCommands {
    RESERVED = 0;
    ENABLE_STEPPER_DRIVER = 1;
    DISABLE_STEPPER_DRIVER = 2;
    ENABLE_STEPPER_VBOOST = 3;
    DISABLE_STEPPER_VBOOST = 4;
}

message SendPinchValveCommandMessage {
    uint32 node_id = 1;
    PinchValveCommands command = 2;
}

message SendManyPinchValveCommandMessage {
    map<uint32, PinchValveCommands> commands = 1;
}
